File: sw/rust/sec/src/legacy.rs
//! Legacy security pattern.
//!
//! The authority is an ACL-style registry of attestations per principal.
//! Revocation removes the scope immediately. The integrity seal binds only
//! the record and the path; the principal and the epoch do not participate.
//! This is the baseline that the tagma-sec pattern is compared against in
//! the shared workflow tests.

use alloc::collections::BTreeMap;
use alloc::vec::Vec;

use crate::authority::Authority;
use crate::integrity::Integrity;
use crate::types::{Action, Attestation, Decision, Epoch, Path, PrincipalId, Scope, Seal};

const LEGACY_INTEGRITY_KEY: &[u8; 32] = b"tagma-sec-poc-legacy-int-key-000";

/// ACL-style authority for the legacy pattern.
pub struct LegacyAuthority {
    next_id: u64,
    next_att: u64,
    attestations: BTreeMap<PrincipalId, Vec<Attestation>>,
}

impl LegacyAuthority {
    pub fn new() -> Self {
        Self {
            next_id: 0,
            next_att: 0,
            attestations: BTreeMap::new(),
        }
    }
}

impl Default for LegacyAuthority {
    fn default() -> Self {
        Self::new()
    }
}

impl Authority for LegacyAuthority {
    fn register(&mut self, _cred: &[u8]) -> PrincipalId {
        self.next_id += 1;
        self.next_id
    }

    fn issue(
        &mut self,
        principal: PrincipalId,
        scope: Scope,
        issued_epoch: Epoch,
        valid_until: Epoch,
    ) -> Option<Attestation> {
        self.next_att += 1;
        let att = Attestation {
            id: self.next_att,
            principal,
            scope,
            issued_epoch,
            valid_until,
        };
        self.attestations
            .entry(principal)
            .or_default()
            .push(att.clone());
        Some(att)
    }

    fn authorize(&self, att: &Attestation, path: &Path, _action: Action, epoch: Epoch) -> Decision {
        let allowed = self.attestations.get(&att.principal).is_some_and(|list| {
            list.iter().any(|a| {
                a.id == att.id
                    && a.scope.matches(path)
                    && epoch >= a.issued_epoch
                    && epoch <= a.valid_until
            })
        });
        if allowed {
            Decision::Allow
        } else {
            Decision::Deny
        }
    }

    fn revoke(&mut self, principal: PrincipalId, scope: &Scope, _epoch: Epoch) {
        if let Some(list) = self.attestations.get_mut(&principal) {
            list.retain(|a| a.scope != *scope);
        }
    }
}

/// Integrity seal for the legacy pattern: binds record and path only.
pub struct LegacyIntegrity;

impl Integrity for LegacyIntegrity {
    fn seal(&self, record: &[u8], path: &Path, _principal: PrincipalId, _epoch: Epoch) -> Seal {
        let mut h = blake3::Hasher::new_keyed(LEGACY_INTEGRITY_KEY);
        h.update(record);
        for c in path {
            h.update(&c.index().to_le_bytes());
        }
        Seal {
            tag: h.finalize().into(),
        }
    }

    fn verify(
        &self,
        record: &[u8],
        path: &Path,
        _principal: PrincipalId,
        _epoch: Epoch,
        seal: &Seal,
    ) -> bool {
        self.seal(record, path, 0, 0).tag == seal.tag
    }

    fn refresh(
        &self,
        record: &[u8],
        path: &Path,
        principal: PrincipalId,
        _from_epoch: Epoch,
        to_epoch: Epoch,
        seal: &Seal,
    ) -> Option<Seal> {
        if self.verify(record, path, principal, 0, seal) {
            Some(self.seal(record, path, principal, to_epoch))
        } else {
            None
        }
    }
}

File: sw/cpp/tagma_sec/src/legacy.cpp
// Legacy security pattern. Mirrors sw/rust/sec/src/legacy.rs.

#include "tagma_sec/legacy.h"

#include "tagma_sec/hash.h"

#include <algorithm>

namespace tagma_sec {
namespace {

constexpr std::array<uint8_t, 32> make_key(const char (&s)[33]) {
  std::array<uint8_t, 32> k{};
  for (int i = 0; i < 32; ++i) k[i] = static_cast<uint8_t>(s[i]);
  return k;
}

const auto kLegacyKey = make_key("tagma-sec-poc-legacy-int-key-000");

}  // namespace

PrincipalId LegacyAuthority::register_principal(const Bytes&) { return ++next_id_; }

std::optional<Attestation> LegacyAuthority::issue(PrincipalId principal, Scope scope,
                                                  Epoch issued_epoch,
                                                  Epoch valid_until) {
  Attestation att{++next_att_, principal, std::move(scope), issued_epoch, valid_until};
  attestations_[principal].push_back(att);
  return att;
}

Decision LegacyAuthority::authorize(const Attestation& att, const Path& path, Action,
                                    Epoch epoch) {
  const auto it = attestations_.find(att.principal);
  if (it == attestations_.end()) return Decision::Deny;
  for (const auto& a : it->second) {
    if (a.id == att.id && a.scope.matches(path) && epoch >= a.issued_epoch &&
        epoch <= a.valid_until) {
      return Decision::Allow;
    }
  }
  return Decision::Deny;
}

void LegacyAuthority::revoke(PrincipalId principal, const Scope& scope, Epoch) {
  const auto it = attestations_.find(principal);
  if (it == attestations_.end()) return;
  it->second.erase(
      std::remove_if(it->second.begin(), it->second.end(),
                     [&scope](const Attestation& a) { return a.scope == scope; }),
      it->second.end());
}

Seal LegacyIntegrity::seal(const Bytes& record, const Path& path, PrincipalId,
                           Epoch) {
  std::vector<Bytes> parts;
  parts.push_back(record);
  for (const auto& c : path) parts.push_back(le16(c.index()));
  return Seal{keyed_tag(kLegacyKey, parts)};
}

bool LegacyIntegrity::verify(const Bytes& record, const Path& path, PrincipalId,
                             Epoch, const Seal& seal) {
  return this->seal(record, path, 0, 0).tag == seal.tag;
}

std::optional<Seal> LegacyIntegrity::refresh(const Bytes& record, const Path& path,
                                             PrincipalId principal, Epoch,
                                             Epoch to_epoch, const Seal& seal) {
  if (!this->verify(record, path, principal, 0, seal)) return std::nullopt;
  return this->seal(record, path, principal, to_epoch);
}

}  // namespace tagma_sec
